carp.common/src/commonMain/kotlin/dk/cachet/carp/common/ddd/AggregateRoot.kt

```diff
@@ -22,6 +22,12 @@ abstract class AggregateRoot<TRoot, TSnapshot : Snapshot<TRoot>, TEvent : Domain
      */
     protected fun event( event: TEvent ) = events.add( event )
 
+    /**
+     * In case the instance this is called on is true, [createEvent] and add it to the event queue of this aggregate root.
+     */
+    protected fun Boolean.eventOnSuccess( createEvent: () -> TEvent ): Boolean =
```

Comment:
The fp in me want's an even more generalized approach here that might even be defined in `common`. Some specific downsides to the concrete implementation are 

1. Unclear semantics: the function returns a boolean but it's always `this` re-emitted.
2. It takes a function as argument that returns an event type, but it is never used.

My suggestion is to add extensions to `Boolean` such as `andThen(block: () -> Any):  Boolean` where the return value must me provided at the call. This would make sense as you sometimes use the returned boolean, sometimes discard it.

Reply:
1. The returned `Boolean` is never used, but it _could_ be useful for chaining, e.g.:

```
doSomething()
    .eventOnSuccess( Event.SomethingSuccess() )
    .eventOnFailure( Event.SomethingFailed() )
```

2. It does not return an event type, it returns an _event_. This is the main point of this helper; the event which is created is emitted.

Does this make it clearer? Any suggestions to make the semantics clearer? The suggestion doesn't seem to match the current behavior. A more generic solution would be something akin to: `doSomething().onSuccess( event( Event.SomethingSuccess() ) )` which is more verbose.

Reply:
ad. 1 You're using the return value in multiple places in carp.protocols.core/src/commonMain/kotlin/dk/cachet/carp/protocols/domain/StudyProtocol.kt
ad. 2. I don't see I was that ambiguous here. It returns something of an event type == an event. Just saying this could be extended to all types, as a general concept of running a block of code _if_ the value is `true`

Perhaps putting this as an extension for Boolean was a bit of a stretch in the first place. If it should be focused on this very specific action, why not a helper function in the `AggregateRoot`

Reply:
1. Correct. Overlooked those cases.

2. I'm still unclear on the alternative you suggest or what that would look like. If not a (protected, thus only accessible by `AggregateRoot`) extension for Boolean, I think it would necessarily involve re-introducing the `if` checks this helper method abstracts away? https://github.com/cph-cachet/carp.core-kotlin/pull/201/commits/34c0deda80a8e9716bf76ecb19e630d3e693d9b8

Reply:
Maybe a more generalizable abstraction is `eventOn( objectToCompare, createEvent )`? Thus, `.eventOn( true ) { Event.SomethingSuccess() }`